rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,526 @@
+
+interface BroadcastOptions {
+  /**
+   * If true, will also send the message to the worker that sent it.
+   * Defaults to false.
+   */
+  sendToSelf?: boolean;
+  /**
+   * If true, the message will also be sent to subscribers subscribing after the message was effectively sent.
+   * Defaults to false.
+   */
+  persist?: boolean;
+}
+
+export interface ClusteringServiceSetup {
+  /**
+   * Return true if clustering mode is enabled, false otherwise
+   */
+  isEnabled: () => boolean;
+  /**
+   * Return the current worker's id. In non-clustered mode, will return `1`
+   */
+  getWorkerId: () => number;
+  /**
+   * Broadcast a message to other workers.
+   * In non-clustered mode, this is a no-op
+   */
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  /**
+   * Registered a handler for given `type` of IPC messages
+   * In non-clustered mode, this is a no-op that returns a no-op unsubscription callback.
+   */
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+  /**
+   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will return true
+   */
+  isMainWorker: () => boolean;
+}
+```
+
+## Cross-worker communication
+
+For some of our changes (such as the /status API, see below), we will need some kind of cross-worker communication. This 
+will need to pass through the coordinator, which will also serve as an 'event bus', or IPC forwarder.
+
+This IPC API will be exposed from the clustering service.
+
+```ts
+export interface ClusteringServiceSetup {
+  // [...]
+  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
+  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
+}
+```
+
+Notes:
+-  to reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
+   It will avoid to force (most) code to check in which mode Kibana is running before calling them.
+-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.
+
+## Executing code on a single worker
+
+In some scenarios, we would like to have parts of the code executed only from a single process. 
+
+The SO migration would be a good example: 
+we don't need to have each worker try to perform the migration, and we'd prefer to have one performing/trying the migration, 
+and the others wait for it. Due to the architecture, we can't have the coordinator perform such single-process jobs, 
+as it doesn't actually run a Kibana server.
+
+There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 
+'main worker'. The coordinator would arbitrary elect a worker as the 'main' one at startup. The clustering service would then expose 
+an API to let workers identify themselves as main or not.
```

Comment:
IMO, a main worker per plugin seems like a high cost of complexity. I think it's much simpler to reason about there being a single main process rather than distributing this responsibility across the workers on a per-plugin basis.

I hope that we can find a way for things like Task Manager and Telemetry to operate correctly on multiple workers within the cluster, just as it does today on multiple instances.

Reply:
> IMO, a main worker per plugin seems like a high cost of complexity

Yea, also, we can't be sure that code executed by the 'main' worker from a given plugin doesn't need something done by the same main worker by another plugin (even if that seems terrible design).

> Otherwise, it feels like clustering would only help with HTTP requests because all the other background processes will likely require one 'main' node for the sake of avoiding concurrency issues.

In the initial implementation, maybe. Even if I don't think background tasks are really a significant part of our load, especially under heavy load (which is what clustering want to address).

More longer term, plugins should adapt their implementation to add inter-worker communication. But we can't realistically have all this done in the initial implem, which is what this main worker concept tries to address.